common/tools/dev-tool/src/framework/command.ts

```diff
@@ -0,0 +1,141 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license
+
+import { CommandLoader } from "./CommandModule";
+import { createPrinter } from "../util/printer";
+import { printCommandUsage, commandStack } from "./printCommandUsage";
+import { ParsedOptions, parseOptions } from "./parseOptions";
+import { CommandInfo, CommandOptions } from "./CommandInfo";
+
+/**
+ * Utility type that makes the type of the "allowMultiple" key in
+ * any options descriptions strictly "undefined" when it is not
+ * specified in the context of a refined extending type
+ */
+export type StrictAllowMultiple<Opts extends CommandOptions> = {
+  [K in keyof Opts]: boolean extends Opts[K]["allowMultiple"]
+    ? Opts[K] & { allowMultiple?: undefined }
+    : Opts[K];
+};
+
+/**
+ * Create a CommandInfo object that describes a command and its functionality.
+ *
+ * For the best type-checking support, please ensure that the `options` parameter
+ * is given to this function as a literal.
+ *
+ * Example:
+ *
+ * ```typescript
+ * export const commandInfo = describe(
+ *   "resolve",
+ *   "display information about the project that owns a directory",
+ *   {
+ *     directory: {
+ *       shortName: "d",
+ *       kind: "string",
+ *       description: "base directory for resolution (uses CWD if unset)",
+ *       allowMultiple: true
+ *     },
+ *     quiet: {
+ *       shortName: "q",
+ *       kind: "boolean",
+ *       default: false,
+ *       description: "output only the directory name with no extra formatting"
+ *     }
+ *   }
+ * );
+ * ```
+ *
+ * @param name the name of this command
+ * @param description a one-line description of this command's functionality
+ * @param options the command's command-line options descriptions (see: {@link CommandOptions})
+ */
+export function makeCommandInfo<Opts extends CommandOptions>(
+  name: string,
+  description: string,
+  options?: Opts
+): CommandInfo<StrictAllowMultiple<Opts>> {
+  return {
+    name,
+    description,
+    options: options as StrictAllowMultiple<Opts>
+  };
+}
+
+/**
+ * Create a subcommand handler that will delegate handling
+ * to a set of lower-order subcommands.
+ *
+ * @param info a the CommandInfo object that describes this command
+ * @param commands map from subcommand name to module implementing that command
+ */
+export function subCommand<Info extends CommandInfo<CommandOptions>>(
+  info: Info,
+  commands: CommandLoader
+): (...args: string[]) => Promise<boolean> {
+  const log = createPrinter(info.name);
+  return async (...rawArgs: string[]) => {
+    commandStack.push(info.name);
```

Comment:
this feels like it might be easy to forget. Not sure if there is a way around it though.

Reply:
I think it would be ideal to be able to have the command stack be created when the program runs as part of the execution of the program instead of trusting that a commandStack will be available in memory. It will make it less likely to run into unintended bugs later. Think of what Jeff is mentioning, it could be hard to remember what is using this as this changes (and software always changes).